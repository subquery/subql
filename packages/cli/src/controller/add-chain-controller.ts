// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

import * as fs from 'fs';
import * as path from 'path';
import {DEFAULT_MULTICHAIN_MANIFEST, getProjectRootAndManifest, getSchemaPath, loadFromJsonOrYaml} from '@subql/common';
import {MultichainProjectManifest, ProjectManifestV1_0_0} from '@subql/types-core';
import {Scalar, Document, parseDocument, YAMLSeq, YAMLMap} from 'yaml';

const nodeToDockerImage: Record<string, string> = {
  '@subql/node': 'onfinality/subql-node',
  '@subql/node-ethereum': 'onfinality/subql-node-ethereum',
  '@subql/node-cosmos': 'onfinality/subql-node-cosmos',
  '@subql/node-algorand': 'onfinality/subql-node-algorand',
  '@subql/node-near': 'onfinality/subql-node-near',
  '@subql/node-stellar': 'subquerynetwork/subql-node-stellar',
  '@subql/node-concordium': 'subquerynetwork/subql-node-concordium',
  '@subql/node-starknet': 'subquerynetwork/subql-node-starknet',
};

type DockerComposeDependsOn = {
  condition: string;
};

type DockerComposeEnvironment = {
  [key: string]: string;
};

type DockerComposeHealthcheck = {
  test: Array<string | 'CMD'>;
  interval: string;
  timeout: string;
  retries: number;
};

type DockerComposeService = {
  image?: string;
  depends_on?: Record<string, DockerComposeDependsOn>;
  restart?: string;
  environment?: DockerComposeEnvironment;
  volumes?: string[];
  command?: string[];
  healthcheck?: DockerComposeHealthcheck;
};

export async function addChain(multichain: string, chainManifestPath: string): Promise<void> {
  const multichainManifestPath = determineMultichainManifestPath(multichain);
  const multichainManifest = loadMultichainManifest(multichainManifestPath);
  chainManifestPath = handleChainManifestOrId(chainManifestPath);
  validateAndAddChainManifest(path.parse(multichainManifestPath).dir, chainManifestPath, multichainManifest);
  fs.writeFileSync(multichainManifestPath, multichainManifest.toString());
  await updateDockerCompose(path.parse(multichainManifestPath).dir, chainManifestPath);
}

export function determineMultichainManifestPath(multichain: string): string {
  if (!multichain) {
    throw new Error(`Multichain project path -f not provided`);
  }

  let multichainPath: string;
  if (fs.lstatSync(multichain).isDirectory()) {
    multichainPath = path.resolve(multichain, DEFAULT_MULTICHAIN_MANIFEST);
  } else {
    if (!fs.existsSync(multichain)) {
      throw new Error(`Could not resolve multichain project path: ${multichain}`);
    }
    multichainPath = multichain;
  }

  return multichainPath;
}

export function loadMultichainManifest(multichainManifestPath: string): Document {
  let multichainManifest: Document;

  if (!isMultiChainProject(multichainManifestPath)) {
    throw new Error(`${multichainManifestPath} is an invalid multichain project manifest`);
  }

  if (fs.existsSync(multichainManifestPath)) {
    const content = fs.readFileSync(multichainManifestPath, 'utf8');
    multichainManifest = parseDocument(content);
  } else {
    throw new Error(`Multichain project ${multichainManifestPath} does not exist`);
  }

  return multichainManifest;
}

export function handleChainManifestOrId(chainManifestPath: string): string {
  if (!chainManifestPath) {
    throw new Error('You must provide chain manifest path');
  }

  // Check if the provided chain manifest path exists
  if (!fs.existsSync(chainManifestPath)) {
    throw new Error(`Chain manifest file does not exist at the specified location: ${chainManifestPath}`);
  }

  return path.resolve(chainManifestPath);
}

export function validateAndAddChainManifest(
  projectDir: string,
  chainManifestPath: string,
  multichainManifest: Document
): void {
  // Validate schema paths
  const chainManifestProject = getProjectRootAndManifest(path.resolve(projectDir, chainManifestPath));
  const chainManifestSchemaPath = getSchemaPath(chainManifestProject.root, chainManifestProject.manifests[0]);

  console.log(`Chain manifest: ${chainManifestProject.manifests[0]}`);

  for (const manifestPath of (multichainManifest.get('projects') as YAMLSeq).items.map(
    (item) => (item as Scalar).value as string
  )) {
    const project = getProjectRootAndManifest(path.resolve(projectDir, manifestPath));
    const schemaPath = getSchemaPath(project.root, project.manifests[0]);

    if (schemaPath !== chainManifestSchemaPath) {
      console.error(
        `Error: Schema path in the provided chain manifest is different from the schema path in the existing chain manifest for project: ${project.root}`
      );
      throw new Error('Schema path mismatch error');
    }
  }

  for (const project of (multichainManifest.get('projects') as YAMLSeq<string>).toJSON()) {
    if (path.resolve(projectDir, project as string) === chainManifestPath) {
      console.log(`project ${project} already exists in multichain manifest, skipping addition`);
      return;
    }
  }

  // Add the chain manifest path to multichain manifest
  const relativePath = path.relative(projectDir, chainManifestPath);

  (multichainManifest.get('projects') as YAMLSeq).add(relativePath);
  console.log(`Successfully added chain manifest path: ${relativePath}`);
}

export function loadDockerComposeFile(dockerComposePath: string): Document | undefined {
  if (fs.existsSync(dockerComposePath)) {
    const content = fs.readFileSync(dockerComposePath, 'utf8');
    return parseDocument(content);
  }
}

function getSubqlNodeService(dockerCompose: Document): DockerComposeService | undefined {
  const services = dockerCompose.get('services');
  if (services && services instanceof YAMLMap) {
    const service = services.items.find((item) => {
      const image = item.value.get('image') as string;
      return image?.startsWith('onfinality/subql-node');
    });

    if (service && service.value instanceof YAMLMap) {
      const commands = service.value.get('command') as YAMLSeq<string>;
      if (!(commands.toJSON() as string[]).includes('--multi-chain')) {
        commands.add('--multi-chain');
      }
      return service.value.toJSON() as DockerComposeService;
    }
  }

  return undefined;
}

async function getDefaultServiceConfiguration(
  chainManifestPath: string,
  serviceName: string
): Promise<DockerComposeService> {
  const manifest = await loadFromJsonOrYaml(chainManifestPath);
  let nodeName: string;
  try {
    nodeName = (manifest as ProjectManifestV1_0_0).runner.node.name;
  } catch (e) {
    throw new Error(`unable to retrieve runner node from manifest: ${e}`);
  }

  const dockerImage = nodeToDockerImage[nodeName];
  if (!dockerImage) {
    throw new Error(`unknown node runner name ${nodeName}`);
  }

  return {
    image: `${dockerImage}:latest`,
    depends_on: {
      postgres: {
        condition: 'service_healthy',
      },
    },
    restart: 'always',
    environment: {
      DB_USER: 'postgres',
      DB_PASS: 'postgres',
      DB_DATABASE: 'postgres',
      DB_HOST: 'postgres',
      DB_PORT: '5432',
    },
    volumes: ['./:/app'],
    command: [`-f=app/${path.basename(chainManifestPath)}`, '--multi-chain'],
    healthcheck: {
      test: ['CMD', 'curl', '-f', `http://${serviceName}:3000/ready`],
      interval: '3s',
      timeout: '5s',
      retries: 10,
    },
  };
}

export async function updateDockerCompose(projectDir: string, chainManifestPath: string): Promise<void> {
  const serviceName = `subquery-node-${path.basename(chainManifestPath, '.yaml')}`;
  const dockerComposePath = path.join(projectDir, 'docker-compose.yml');
  const dockerCompose = loadDockerComposeFile(dockerComposePath);
  if (!dockerCompose) {
    throw new Error(`Docker Compose file does not exist at the specified location: ${dockerComposePath}`);
  }

  console.log(`Updating Docker Compose for chain manifest: ${chainManifestPath}`);

  //check if service already exists
  const services = dockerCompose.get('services');
  if (services && services instanceof YAMLMap) {
    const existingService = services.items.find((item) => {
      const image = item.value.toJSON().image;
      if (!image || !image.startsWith('onfinality/subql-node')) {
        return false;
      }
      const commands: string[] = item.value.toJSON().command;
      return commands?.includes(`-f=app/${path.basename(chainManifestPath)}`);
    });

    if (existingService) {
      console.log(
        `Service for ${path.basename(chainManifestPath)} already exists in Docker Compose, skipping addition`
      );
      return;
    }
  }

  let subqlNodeService = getSubqlNodeService(dockerCompose);
  if (subqlNodeService) {
    // If the service already exists, update its configuration
    subqlNodeService.command = subqlNodeService.command?.filter((cmd) => !cmd.startsWith('-f=')) ?? [];
    subqlNodeService.command.push(`-f=app/${path.basename(chainManifestPath)}`);

    if (subqlNodeService.healthcheck) {
      subqlNodeService.healthcheck.test = ['CMD', 'curl', '-f', `http://${serviceName}:3000/ready`];
    }
  } else {
    // Otherwise, create a new service configuration
    subqlNodeService = await getDefaultServiceConfiguration(chainManifestPath, serviceName);
  }
  console.log(`Created new service configuration: ${serviceName}`);

  (services as YAMLMap).add({key: serviceName, value: subqlNodeService});

  // Save the updated Docker Compose file
  fs.writeFileSync(dockerComposePath, dockerCompose.toString());
  console.log(`Docker Compose file updated successfully at: ${dockerComposePath}`);
}

function isMultiChainProject(content: string): boolean {
  return Array.isArray((loadFromJsonOrYaml(content) as MultichainProjectManifest).projects);
}
