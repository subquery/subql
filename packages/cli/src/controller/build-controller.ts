// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import path from 'path';
import webpack from 'webpack';
import {merge} from 'webpack-merge';

const getBaseConfig = (entryPath: string, outputPath: string, development?: boolean): webpack.Configuration => ({
  target: 'node',
  mode: development ? 'development' : 'production',
  entry: entryPath,
  devtool: development && 'inline-source-map',
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        exclude: /node_modules/,
        loader: require.resolve('ts-loader'),
        options: {
          compilerOptions: {
            declaration: false,
          },
        },
      },
      {
        test: /\.ya?ml$/,
        use: 'yaml-loader',
      },
    ],
  },

  resolve: {
    extensions: ['.tsx', '.ts', '.js', '.json'],
  },

  output: {
    path: path.dirname(outputPath),
    filename: path.basename(outputPath),
    libraryTarget: 'commonjs',
  },
});

export async function runWebpack(entryPath: string, outputPath: string, isDev = false, clean = false): Promise<void> {
  const config = merge(
    getBaseConfig(entryPath, outputPath, isDev),
    {output: {clean}}
    // Can allow projects to override webpack config here
  );

  await new Promise((resolve, reject) => {
    webpack(config).run((error, stats) => {
      if (error) {
        reject(error);
        return;
      }

      if (stats.hasErrors()) {
        const info = stats.toJson();

        reject(info.errors[0]);
        this.log(info.errors[0].details);
        return;
      }

      resolve(true);
    });
  });
}
