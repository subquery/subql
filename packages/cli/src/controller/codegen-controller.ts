// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import fs from 'fs';
import path from 'path';
import {promisify} from 'util';
import {
  getAllEntitiesRelations,
  getAllJsonObjects,
  setJsonObjectType,
  getTypeByScalarName,
  GraphQLEntityField,
  GraphQLJsonFieldType,
  GraphQLEntityIndex,
  getAllEnums,
  loadProjectManifest,
  ProjectManifestVersioned,
  isCustomDs,
} from '@subql/common';
import ejs from 'ejs';
import {upperFirst, uniq} from 'lodash';
import rimraf from 'rimraf';

const MODEL_TEMPLATE_PATH = path.resolve(__dirname, '../template/model.ts.ejs');
const MODELS_INDEX_TEMPLATE_PATH = path.resolve(__dirname, '../template/models-index.ts.ejs');
const TYPES_INDEX_TEMPLATE_PATH = path.resolve(__dirname, '../template/types-index.ts.ejs');
const INTERFACE_TEMPLATE_PATH = path.resolve(__dirname, '../template/interface.ts.ejs');
const ENUM_TEMPLATE_PATH = path.resolve(__dirname, '../template/enum.ts.ejs');
const DYNAMIC_DATASOURCE_TEMPLATE_PATH = path.resolve(__dirname, '../template/datasource-templates.ts.ejs');
const TYPE_ROOT_DIR = 'src/types';
const MODEL_ROOT_DIR = 'src/types/models';
const exportTypes = {
  models: false,
  interfaces: false,
  enums: false,
  datasources: false,
};

// 4. Render entity data in ejs template and write it
export async function renderTemplate(templatePath: string, outputPath: string, templateData: ejs.Data): Promise<void> {
  const data = await ejs.renderFile(templatePath, templateData);
  await fs.promises.writeFile(outputPath, data);
}

// 3. Re-format the field of the entity
export interface ProcessedField {
  name: string;
  type: string;
  required: boolean;
  isEnum: boolean;
  indexed: boolean;
  unique?: boolean;
  isArray: boolean;
  isJsonInterface: boolean;
}

export async function generateJsonInterfaces(projectPath: string, schema: string): Promise<void> {
  const typesDir = path.join(projectPath, TYPE_ROOT_DIR);
  const jsonObjects = getAllJsonObjects(schema);
  const jsonInterfaces = jsonObjects.map((r) => {
    const object = setJsonObjectType(r, jsonObjects);
    const fields = processFields('jsonField', object.name, object.fields);
    return {
      interfaceName: object.name,
      fields,
    };
  });

  if (jsonInterfaces.length !== 0) {
    const interfaceTemplate = {
      props: {
        jsonInterfaces,
      },
      helper: {
        upperFirst,
      },
    };
    try {
      await renderTemplate(INTERFACE_TEMPLATE_PATH, path.join(typesDir, `interfaces.ts`), interfaceTemplate);
      exportTypes.interfaces = true;
    } catch (e) {
      throw new Error(`When render json interfaces having problems.`);
    }
  }
}

export async function generateEnums(projectPath: string, schema: string): Promise<void> {
  const typesDir = path.join(projectPath, TYPE_ROOT_DIR);
  const jsonObjects = getAllEnums(schema);
  const enums = jsonObjects.map((r) => {
    return {
      name: r.name,
      values: r.getValues().map((v) => v.name),
    };
  });

  if (enums.length !== 0) {
    const enumsTemplate = {
      props: {
        enums,
      },
    };
    try {
      await renderTemplate(ENUM_TEMPLATE_PATH, path.join(typesDir, `enums.ts`), enumsTemplate);
      exportTypes.enums = true;
    } catch (e) {
      throw new Error(`When render enums having problems.`);
    }
  }
}

export function processFields(
  type: 'entity' | 'jsonField',
  className: string,
  fields: (GraphQLEntityField | GraphQLJsonFieldType)[],
  indexFields: GraphQLEntityIndex[] = []
): ProcessedField[] {
  const fieldList: ProcessedField[] = [];
  for (const field of fields) {
    const injectField = {
      name: field.name,
      required: !field.nullable,
      isArray: field.isArray,
      isEnum: false,
    } as ProcessedField;
    if (type === 'entity') {
      const [indexed, unique] = indexFields.reduce<[boolean, boolean]>(
        (acc, indexField) => {
          if (indexField.fields.includes(field.name)) {
            acc[0] = true;
            if (indexField.fields.length === 1 && indexField.unique) {
              acc[1] = true;
            } else if (indexField.unique === undefined) {
              acc[1] = false;
            }
          }
          return acc;
        },
        [false, undefined]
      );
      injectField.indexed = indexed;
      injectField.unique = unique;
    }
    if ((field as GraphQLEntityField).isEnum) {
      injectField.type = field.type;
      injectField.isEnum = true;
      injectField.isJsonInterface = false;
    } else {
      switch (field.type) {
        default: {
          injectField.type = getTypeByScalarName(field.type).tsType;
          if (!injectField.type) {
            throw new Error(
              `Schema: undefined type "${field.type.toString()}" on field "${
                field.name
              }" in "type ${className} @${type}"`
            );
          }
          injectField.isJsonInterface = false;
          break;
        }
        case 'Json': {
          if (field.jsonInterface === undefined) {
            throw new Error(`On field ${field.name} type is Json but json interface is not defined`);
          }
          injectField.type = upperFirst(field.jsonInterface.name);
          injectField.isJsonInterface = true;
        }
      }
    }
    fieldList.push(injectField);
  }
  return fieldList;
}

async function prepareDirPath(path: string, recreate: boolean) {
  try {
    await promisify(rimraf)(path);
    if (recreate) {
      await fs.promises.mkdir(path, {recursive: true});
    }
  } catch (e) {
    throw new Error(`Failed to prepare ${path}`);
  }
}

//1. Prepare models directory and load schema
export async function codegen(projectPath: string): Promise<void> {
  const modelDir = path.join(projectPath, MODEL_ROOT_DIR);
  const interfacesPath = path.join(projectPath, TYPE_ROOT_DIR, `interfaces.ts`);
  await prepareDirPath(modelDir, true);
  await prepareDirPath(interfacesPath, false);

  const manifest = loadProjectManifest(projectPath);

  await generateJsonInterfaces(projectPath, path.join(projectPath, manifest.schema));
  await generateModels(projectPath, path.join(projectPath, manifest.schema));
  await generateEnums(projectPath, path.join(projectPath, manifest.schema));
  await generateDatasourceTemplates(projectPath, manifest);
  if (exportTypes.interfaces || exportTypes.models || exportTypes.enums || exportTypes.datasources) {
    try {
      await renderTemplate(TYPES_INDEX_TEMPLATE_PATH, path.join(projectPath, TYPE_ROOT_DIR, `index.ts`), {
        props: {
          exportTypes,
        },
      });
    } catch (e) {
      throw new Error(`When render index in types having problems.`);
    }
    console.log(`* Types index generated !`);
  }
}

// 2. Loop all entities and render it
export async function generateModels(projectPath: string, schema: string): Promise<void> {
  const extractEntities = getAllEntitiesRelations(schema);
  for (const entity of extractEntities.models) {
    const baseFolderPath = '.../../base';
    const className = upperFirst(entity.name);
    const entityName = entity.name;
    const fields = processFields('entity', className, entity.fields, entity.indexes);
    const importJsonInterfaces = uniq(fields.filter((field) => field.isJsonInterface).map((f) => f.type));
    const importEnums = fields.filter((field) => field.isEnum).map((f) => f.type);
    const indexedFields = fields.filter((field) => field.indexed && !field.isJsonInterface);
    const modelTemplate = {
      props: {
        baseFolderPath,
        className,
        entityName,
        fields,
        importJsonInterfaces,
        importEnums,
        indexedFields,
      },
      helper: {
        upperFirst,
      },
    };
    try {
      await renderTemplate(
        MODEL_TEMPLATE_PATH,
        path.join(projectPath, MODEL_ROOT_DIR, `${className}.ts`),
        modelTemplate
      );
    } catch (e) {
      console.error(e);
      throw new Error(`When render entity ${className} to schema having problems.`);
    }
    console.log(`* Schema ${className} generated !`);
  }
  const classNames = extractEntities.models.map((entity) => entity.name);
  if (classNames.length !== 0) {
    try {
      await renderTemplate(MODELS_INDEX_TEMPLATE_PATH, path.join(projectPath, MODEL_ROOT_DIR, `index.ts`), {
        props: {
          classNames,
        },
        helper: {
          upperFirst,
        },
      });
      exportTypes.models = true;
    } catch (e) {
      throw new Error(`When render index in models having problems.`);
    }
    console.log(`* Models index generated !`);
  }
}

export async function generateDatasourceTemplates(
  projectPath: string,
  projectManifest: ProjectManifestVersioned
): Promise<void> {
  if (!projectManifest.isV0_2_1) return;

  const manifest = projectManifest.asV0_2_1;

  if (!manifest.templates?.length) return;

  try {
    const props = manifest.templates.map((t) => ({
      name: t.name,
      args: isCustomDs(t) ? 'Record<string, unknown>' : undefined,
    }));
    await renderTemplate(DYNAMIC_DATASOURCE_TEMPLATE_PATH, path.join(projectPath, TYPE_ROOT_DIR, `datasources.ts`), {
      props,
    });

    exportTypes.datasources = true;
  } catch (e) {
    console.error(e);
    throw new Error(`Unable to generate datasource template constructors`);
  }
  console.log(`* Datasource template constructors generated !`);
}
