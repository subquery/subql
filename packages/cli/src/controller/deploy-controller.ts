// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import axios from 'axios';
import {deploymentDataType, validateDataType} from '../types';
import {errorHandle} from '../utils';

export async function deployToHostedService(
  org: string,
  project_name: string,
  authToken: string,
  ipfsCID: string,
  indexerImageVersion: string,
  queryImageVersion: string,
  endpoint: string,
  type: string,
  dictEndpoint: string,
  url: string
): Promise<deploymentDataType> {
  const key = `${encodeURIComponent(org)}/${encodeURIComponent(project_name)}`;
  try {
    const result = (
      await axios({
        headers: {
          Authorization: `Bearer ${authToken}`,
        },
        method: 'post',
        url: `v2/subqueries/${key}/deployments`,
        baseURL: url,
        data: {
          version: ipfsCID,
          dictEndpoint: dictEndpoint,
          endpoint: endpoint,
          advancedSettings: {
            '@subql/node': {},
            '@subql/query': {},
          },
          indexerImageVersion: indexerImageVersion,
          queryImageVersion: queryImageVersion,
          type: type,
        },
      })
    ).data;
    return result.deployment;
  } catch (e) {
    errorHandle(e, 'Error deploying to hosted service:');
  }
}

export async function promoteDeployment(
  org: string,
  project_name: string,
  authToken: string,
  deploymentId: number,
  url: string
): Promise<string> {
  const key = `${encodeURIComponent(org)}/${encodeURIComponent(project_name)}`;
  try {
    await axios({
      headers: {
        Authorization: `Bearer ${authToken}`,
      },
      method: 'post',
      url: `subqueries/${key}/deployments/${deploymentId}/release`,
      baseURL: url,
    });
    return `${deploymentId}`;
  } catch (e) {
    errorHandle(e, 'Failed to promote project:');
  }
}

export async function deleteDeployment(
  org: string,
  project_name: string,
  authToken: string,
  deploymentId: number,
  url: string
): Promise<string> {
  const key = `${encodeURIComponent(org)}/${encodeURIComponent(project_name)}`;
  try {
    await axios({
      headers: {
        Authorization: `Bearer ${authToken}`,
      },
      method: 'delete',
      url: `subqueries/${key}/deployments/${deploymentId}`,
      baseURL: url,
    });
    return `${deploymentId}`;
  } catch (e) {
    errorHandle(e, 'Failed to delete deployment:');
  }
}

export async function deploymentStatus(
  org: string,
  project_name: string,
  authToken: string,
  deployID: number,
  url: string
): Promise<string> {
  const key = `${encodeURIComponent(org)}/${encodeURIComponent(project_name)}`;
  try {
    const result = (
      await axios({
        headers: {
          Authorization: `Bearer ${authToken}`,
        },
        method: 'get',
        url: `subqueries/${key}/deployments/${deployID}/status`,
        baseURL: url,
      })
    ).data;
    return `${result.status}`;
  } catch (e) {
    errorHandle(e, 'Failed to get deployment status:');
  }
}

export async function ipfsCID_validate(cid: string, authToken: string, url: string): Promise<validateDataType> {
  try {
    const result = (
      await axios({
        headers: {
          Authorization: `Bearer ${authToken}`,
        },
        method: 'post',
        url: `ipfs/deployment-id/${cid}/validate`,
        baseURL: url,
      })
    ).data;
    return result;
  } catch (e) {
    errorHandle(e, 'Failed to validate IPFS CID:');
  }
}

export async function getDictEndpoints(url: string): Promise<endpointType[]> {
  try {
    const result = (
      await axios({
        method: 'get',
        url: `subqueries/dictionaries`,
        baseURL: url,
      })
    ).data;
    return result;
  } catch (e) {
    errorHandle(e, 'Failed to get dictionary endpoint:');
  }
}

export async function getEndpoints(url: string): Promise<endpointType[]> {
  try {
    const result = (
      await axios({
        method: 'get',
        url: `subqueries/network-endpoints`,
        baseURL: url,
      })
    ).data;
    return result;
  } catch (e) {
    errorHandle(e, 'Failed to get endpoint:');
  }
}

export function processEndpoints(endpoints: endpointType[], chainId: string): string | undefined {
  return endpoints.find((endpoint: endpointType) => endpoint.chainId === chainId)?.endpoint;
}

export async function getImage_v(name: string, version: string, authToken: string, url: string): Promise<string[]> {
  try {
    const result = (
      await axios({
        headers: {Authorization: `Bearer ${authToken}`},
        method: 'get',
        url: `info/images/${encodeURIComponent(name)}?version=${encodeURIComponent(version)}`,
        baseURL: url,
      })
    ).data;
    return result;
  } catch (e) {
    errorHandle(e, 'Failed to get image:');
  }
}
interface endpointType {
  network: string;
  endpoint: string;
  chainId: string;
}
