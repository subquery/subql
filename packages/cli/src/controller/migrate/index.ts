// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

export * from './migrate-controller.js';
export * from './manifest/migrate-manifest.controller.js';
export * from './abis/migrate-abis.controller.js';
export * from './schema/migrate-schema.controller.js';
export * from './constants.js';
