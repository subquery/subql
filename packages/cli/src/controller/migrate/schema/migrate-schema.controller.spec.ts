// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

import {migrateSchemaFromString} from './migrate-schema.controller.js';

describe('Subgraph Graphql Schema migration', () => {
  it('correctly migrates a schema', () => {
    // This schema is a mix of schemas from https://github.com/graphprotocol/graph-tooling/tree/main/examples as well as the docs
    const schema = `
    type _Schema_
    @fulltext(
      name: "exampleSearch"
      language: en
      algorithm: rank
      include: [{ entity: "ExampleEntity", fields: [{ name: "optionalString" }, { name: "requiredString" }, { name: "optionalStringList" }] }]
    )

    """
    The Example entity has an example of various fields
    """
    type ExampleEntity @entity {
      """
      The id is required and a unique identifier of the entity
      """
      id: ID!

      optionalBoolean: Boolean
      requiredBoolean: Boolean!
      optionalBooleanList: [Boolean!]
      requiredBooleanList: [Boolean!]!

      optionalString: String
      requiredString: String!
      optionalStringList: [String!]
      requiredStringList: [String!]!

      optionalBytes: Bytes
      requiredBytes: Bytes!
      optionalBytesList: [Bytes!]
      requiredBytesList: [Bytes!]!

      optionalInt: Int
      requiredInt: Int!
      optionalIntList: [Int!]
      requiredIntList: [Int!]!

      optionalInt8: Int8
      requiredInt8: Int8!
      optionalInt8List: [Int8!]
      requiredInt8List: [Int8!]!

      optionalBigInt: BigInt
      requiredBigInt: BigInt!
      optionalBigIntList: [BigInt!]
      requiredBigIntList: [BigInt!]!

      optionalBigDecimal: BigDecimal
      requiredBigDecimal: BigDecimal!
      optionalBigDecimalList: [BigDecimal!]
      requiredBigDecimalList: [BigDecimal!]!

      optionalTimestamp: Timestamp
      requiredTimestamp: Timestamp!
      optionalTimestampList: [Timestamp!]
      requiredTimestampList: [Timestamp!]!

      optionalReference: OtherEntity
      requiredReference: OtherEntity!
      optionalReferenceList: [OtherEntity!]
      requiredReferenceList: [OtherEntity!]!

      derivedEntity: [FooEntity!]! @derivedFrom(field: "example")
      derivedEntity2: [FooEntity!]!
    }

    type OtherEntity @entity(immutable: true, timeseries: true) {
      id: ID!
    }

    type FooEntity @entity {
      id: Bytes!

      example: ExampleEntity!
    }

    type Stats @aggregation(intervals: ["hour", "day"], source: "Block") {
      # The id; it is the id of one of the data points that were aggregated into
      # this bucket, but which one is undefined and should not be relied on
      id: Int8!
      # The timestamp of the bucket is always the timestamp of the beginning of
      # the interval
      timestamp: Timestamp!

      # The aggregates

      # A count of the number of data points that went into this bucket
      count: Int! @aggregate(fn: "count")
      # The max(number) of the data points for this bucket
      max: BigDecimal! @aggregate(fn: "max", arg: "number")
      min: BigDecimal! @aggregate(fn: "min", arg: "number")
      # sum_{i=n}^m i = (m - n + 1) * (n + m) / 2
      sum: BigInt! @aggregate(fn: "sum", arg: "number")
      first: Int! @aggregate(fn: "first", arg: "number")
      last: Int! @aggregate(fn: "last", arg: "number")
      maxGas: BigInt! @aggregate(fn: "max", arg: "gasUsed")
      maxDifficulty: BigInt! @aggregate(fn: "max", arg: "difficulty")
    }
    `;

    const migratedSchema = migrateSchemaFromString(schema);

    console.log('Migrated schema', migratedSchema);

    expect(migratedSchema).toMatchSnapshot();
  });
});
