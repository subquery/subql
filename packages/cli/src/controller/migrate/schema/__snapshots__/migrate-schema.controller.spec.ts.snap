// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subgraph Graphql Schema migration correctly migrates a schema 1`] = `
""""The Example entity has an example of various fields"""
type ExampleEntity @entity @fullText(fields: ["optionalString", "requiredString", "optionalStringList"], language: "english") {
  """The id is required and a unique identifier of the entity"""
  id: ID!
  optionalBoolean: Boolean
  requiredBoolean: Boolean!
  optionalBooleanList: [Boolean!]
  requiredBooleanList: [Boolean!]!
  optionalString: String
  requiredString: String!
  optionalStringList: [String!]
  requiredStringList: [String!]!
  optionalBytes: String
  requiredBytes: String!
  optionalBytesList: [String!]
  requiredBytesList: [String!]!
  optionalInt: Int
  requiredInt: Int!
  optionalIntList: [Int!]
  requiredIntList: [Int!]!
  optionalInt8: Int
  requiredInt8: Int!
  optionalInt8List: [Int!]
  requiredInt8List: [Int!]!
  optionalBigInt: BigInt
  requiredBigInt: BigInt!
  optionalBigIntList: [BigInt!]
  requiredBigIntList: [BigInt!]!
  optionalBigDecimal: Float
  requiredBigDecimal: Float!
  optionalBigDecimalList: [Float!]
  requiredBigDecimalList: [Float!]!
  optionalTimestamp: Date
  requiredTimestamp: Date!
  optionalTimestampList: [Date!]
  requiredTimestampList: [Date!]!
  optionalReference: OtherEntity
  requiredReference: OtherEntity!
  optionalReferenceList: [OtherEntity!] @derivedFrom(field: "<replace-me>")
  requiredReferenceList: [OtherEntity!]! @derivedFrom(field: "<replace-me>")
  derivedEntity: [FooEntity!]! @derivedFrom(field: "example")
  derivedEntity2: [FooEntity!]! @derivedFrom(field: "example")
}

type OtherEntity @entity {
  id: ID!
}

type FooEntity @entity {
  id: ID!
  example: ExampleEntity!
}"
`;
